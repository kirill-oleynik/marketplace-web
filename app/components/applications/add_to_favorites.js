import React from 'react';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';

export const AddToFavorites = ({ t, onClick, disabled }) => (
  <button
    type="button"
    onClick={onClick}
    disabled={disabled}
    className="favourites-add"
  >
    <i className="icon icon-heart font-16 mr-10 favourites-heart__icon" />

    <span className="favourites-add__text">
      {t('appProfile.favorites.add')}
    </span>
  </button>
);

AddToFavorites.propTypes = {
  t: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  disabled: PropTypes.bool.isRequired
};

export default translate(['applications'])(AddToFavorites);
