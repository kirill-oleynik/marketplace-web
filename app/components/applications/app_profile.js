import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'reactstrap';
import { translate } from 'react-i18next';

import MainButton from '../main_button';
import ToggleFavorite from './toggle_favorite';
import Rating from '../rating';
import RatingMarks from '../rating_marks';
import ButtonWithIcon from '../button_with_icon';
import ImageGallerySlider from '../image_gallery';
import { asFoundedDate } from '../../helpers/dates_helpers';
import {
  truncateParagraph, humanizeUrl, emailLink, phoneLink
} from '../../helpers/text_helpers';

export class AppProfile extends Component {
  state = {
    descriptionTruncated: true
  };

  setDescriptionTruncated = (flag) => (
    this.setState({
      descriptionTruncated: flag
    })
  );

  getDescriptionLength = () => {
    if (this.descriptionStandardLength()) {
      return this.descriptionLength();
    }

    return this.state.descriptionTruncated ? 500 : this.descriptionLength();
  }

  descriptionLength = () => {
    const { appProfile: { description } } = this.props;
    return description && description.length;
  };

  descriptionStandardLength = () => this.descriptionLength() < 500;

  needToShowMoreDescription = () => (
    this.descriptionStandardLength() ? false : this.state.descriptionTruncated
  );

  needToShowLessDescription = () => (
    this.descriptionStandardLength() ? false : !this.state.descriptionTruncated
  );

  showLessDescription = () => this.setDescriptionTruncated(true);
  showMoreDescription = () => this.setDescriptionTruncated(false);

  render() {
    const {
      t, appProfile, canToggleFavorite, addToFavorites, removeFromFavorites
    } = this.props;

    return (
      <div>
        <div className="d-flex flex-sm-wrap-down mb-50">
          <div>
            <span
              className="app-item__img app-item__img--md mr-40"
              style={{ backgroundImage: `url('${appProfile.logo}')` }}
            />
          </div>

          <div
            className="d-flex flex-wrap flex-grow-1 justify-content-between pt-15"
          >
            <div className="mb-30">
              <p className="font-24 font-700 mb-10">
                {appProfile.title}
              </p>

              <Rating className="mb-30" />

              <ToggleFavorite
                application={appProfile}
                addToFavorites={addToFavorites}
                removeFromFavorites={removeFromFavorites}
                disabled={!canToggleFavorite}
              />
            </div>

            <div>
              <MainButton
                color="blue"
                size="lg"
              >
                {t('appProfile.requestDemo')}
              </MainButton>
            </div>
          </div>
        </div>

        <div className="mb-20">
          <p className="font-20 font-700 mb-30">
            {t('appProfile.gallery.title')}
          </p>

          <div className="mb-30">
            <ImageGallerySlider />
          </div>

          <p className="font-20 font-700">
            {t('appProfile.description.title')}
          </p>

          <p className="in-black-050">
            {
              truncateParagraph(
                appProfile.description,
                this.getDescriptionLength()
              )
            }
          </p>

          {
            this.needToShowMoreDescription() ? (
              <ButtonWithIcon
                icon="arrow-down"
                text={t('appProfile.description.readMore')}
                className="mb-10"
                onClick={this.showMoreDescription}
              />
            ) : null
          }

          {
            this.needToShowLessDescription() ? (
              <ButtonWithIcon
                icon="arrow-up"
                text={t('appProfile.description.readLess')}
                className="mb-10"
                onClick={this.showLessDescription}
              />
            ) : null
          }
        </div>

        <div className="divider divider--dark mb-30" />

        <div>
          <p className="font-20 font-700">
            {t('appProfile.ratings.title')}
          </p>

          <div className="d-flex flex-wrap justify-content-between">
            <div className="mr-70">
              <p className="font-16 font-700 mb-20">
                {t('appProfile.ratings.avarage')}
              </p>

              <div className="d-flex mb-30">
                <div className="mr-25">
                  <p className="font-54 lh-1-2 in-black-050 mb-10">
                    4.2
                  </p>

                  <Rating className="rating--sm" />
                </div>

                <div>
                  <RatingMarks />
                </div>
              </div>
            </div>

            <div>
              <p className="font-16 font-700 mb-20">
                {t('appProfile.ratings.submit')}
              </p>

              <div className="rating__stars">
                <i className="rating__stars-item icon icon-star" />
                <i className="rating__stars-item icon icon-star" />
                <i className="rating__stars-item icon icon-star" />
                <i className="rating__stars-item icon icon-star" />
                <i className="rating__stars-item icon icon-star" />
              </div>
            </div>
          </div>
        </div>

        <div className="divider divider--dark mb-30" />

        <div>
          <p className="font-20 font-700">
            {t('appProfile.extra.title')}
          </p>

          <Row className="mb-15">
            {
              appProfile.founded ? (
                <Col xs="12" sm="6" md="4">
                  <p className="font-16 font-700 mb-5">
                    {t('appProfile.extra.founded')}
                  </p>

                  <p className="in-black-050">
                    {asFoundedDate(appProfile.founded)}
                  </p>
                </Col>
              ) : null
            }

            {
              appProfile.email ? (
                <Col xs="12" sm="6" md="4">
                  <p className="font-16 font-700 mb-5">
                    {t('appProfile.extra.email')}
                  </p>

                  <a
                    href={emailLink(appProfile.email)}
                    className="in-black-050"
                  >
                    {appProfile.email}
                  </a>
                </Col>
              ) : null
            }

            {
              appProfile.website ? (
                <Col xs="12" sm="6" md="4">
                  <p className="font-16 font-700 mb-5">
                    {t('appProfile.extra.website')}
                  </p>

                  <a
                    href={appProfile.website}
                    target="_blank"
                    className="in-black-050"
                  >
                    {humanizeUrl(appProfile.website)}
                  </a>
                </Col>
              ) : null
            }

            {
              appProfile.phone ? (
                <Col xs="12" sm="6" md="4">
                  <p className="font-16 font-700 mb-5">
                    {t('appProfile.extra.phone')}
                  </p>

                  <a
                    href={phoneLink(appProfile.phone)}
                    className="in-black-050"
                  >
                    {appProfile.phone}
                  </a>
                </Col>
              ) : null
            }

            {
              appProfile.address ? (
                <Col xs="12" sm="6" md="4">
                  <p className="font-16 font-700 mb-5">
                    {t('appProfile.extra.address')}
                  </p>

                  <p className="in-black-050">
                    {appProfile.address}
                  </p>
                </Col>
              ) : null
            }
          </Row>
        </div>
      </div>
    );
  }
}

AppProfile.propTypes = {
  t: PropTypes.func.isRequired,
  appProfile: PropTypes.object.isRequired,
  addToFavorites: PropTypes.func.isRequired,
  removeFromFavorites: PropTypes.func.isRequired,
  canToggleFavorite: PropTypes.bool.isRequired
};

export default translate(['applications'])(AppProfile);
