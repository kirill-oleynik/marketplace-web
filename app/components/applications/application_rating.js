import React from 'react';
import PropTypes from 'prop-types';
import RatingPreview from './rating_preview';
import { averageRating, votePercentage } from '../../helpers/number_helpers';

export const ApplicationRating = ({ t, data }) => (
  <div className="mr-70">
    <p className="font-16 font-700 mb-20">
      {t('appProfile.ratings.avarage')}
    </p>

    <div className="d-flex mb-30">
      <div className="mr-25">
        <p className="font-54 lh-1-2 in-black-050 mb-10">
          {averageRating(data.average)}
        </p>

        <RatingPreview
          t={t}
          total={data.total}
          average={data.average}
          className="rating--sm"
        />
      </div>

      <div>
        <table
          cellSpacing="0"
          cellPadding="0"
        >
          <tbody>
            {
              [5, 4, 3, 2, 1].map((index) => (
                <tr
                  key={index}
                  className="rating-marks"
                >
                  <td>
                    <span className="font-12 mr-20">
                      {t('appProfile.ratings.votesPoints', { value: index })}
                    </span>
                  </td>

                  <td>
                    <div className="relative">
                      <div className="rating-marks__line" />

                      <div
                        className="rating-marks__line-value"
                        style={{
                          width: votePercentage(data.votes, index)
                        }}
                      />
                    </div>
                  </td>

                  <td>
                    <div className="rating-marks__value">
                      {data.votes[index]}
                    </div>
                  </td>
                </tr>
              ))
            }
          </tbody>
        </table>
      </div>
    </div>
  </div>
);

ApplicationRating.propTypes = {
  t: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired
};

export default ApplicationRating;
