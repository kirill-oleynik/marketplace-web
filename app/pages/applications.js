import React, { Component } from 'react';

import Head from 'next/head';
import PropTypes from 'prop-types';
import { I18nextProvider } from 'react-i18next';

import withReduxAndSaga from '../store';
import withTranslations from '../with_translations';

import { load } from '../actions/page_actions';
import { getAppProfile } from '../selectors/applications_selectors';

import MainLayout from '../layouts/main_layout';
import ApplicationContainer from '../containers/application_container';

class Applications extends Component {
  static propTypes = {
    i18n: PropTypes.object.isRequired,
    application: PropTypes.object.isRequired
  }

  static async getInitialProps({ store, ...rest }) {
    store.dispatch(
      load({
        name: 'applications',
        context: rest
      })
    );
  }

  render() {
    const { i18n, application } = this.props;

    return (
      <I18nextProvider i18n={i18n}>
        <div>
          <Head>
            <title>
              {application.title ? `${application.title} - ` : ''}
              {i18n.t('applications:pageTitle')}
            </title>
          </Head>

          <MainLayout>
            <ApplicationContainer />
          </MainLayout>
        </div>
      </I18nextProvider>
    );
  }
}

const mapStateToProps = (state) => ({
  application: getAppProfile(state)
});

export default withReduxAndSaga(
  withTranslations('applications', 'common')(Applications),
  mapStateToProps
);
