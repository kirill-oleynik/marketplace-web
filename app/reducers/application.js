import { combineReducers } from 'redux';
import { SUCCESS, APPLICATIONS_FETCH_GALLERY } from '../constants';

export const gallery = (state = {}, action = {}) => {
  switch (action.type) {
    case APPLICATIONS_FETCH_GALLERY + SUCCESS:
      return action.payload.gallery;
    default:
      return state;
  }
};

export default combineReducers({
  gallery
});
