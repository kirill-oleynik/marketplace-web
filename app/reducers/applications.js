import { combineReducers } from 'redux';

import {
  SUCCESS, FAILURE, REQUEST, CATEGORIES_FETCH, CATEGORIES_FETCH_ALL,
  APPLICATION_FETCH, APPLICATIONS_ADD_TO_FAVORITES, FAVORITES_FETCH_ALL,
  APPLICATIONS_REMOVE_FROM_FAVORITES, REVIEW_CREATE
} from '../constants';

const applicationsReduceFn = (accumulator, application) => ({
  ...accumulator,
  [application.id]: application
});

const applicationIdReduceFn = (accumulator, application) => ([
  ...accumulator,
  application.id
]);

export const byId = (state = {}, action) => {
  switch (action.type) {
    case CATEGORIES_FETCH + SUCCESS:
      return action.payload.category.applications.reduce(
        applicationsReduceFn, state
      );
    case CATEGORIES_FETCH_ALL + SUCCESS:
      return action.payload.categories.reduce((accumulator, category) => (
        {
          ...accumulator,
          ...category.applications.reduce(applicationsReduceFn, {})
        }
      ), state);
    case FAVORITES_FETCH_ALL + SUCCESS:
      return action.payload.favorites.reduce((accumulator, favorite) => (
        {
          ...accumulator,
          ...[favorite.application].reduce(applicationsReduceFn, {})
        }
      ), state);
    case APPLICATION_FETCH + SUCCESS:
      return {
        ...state,
        [action.payload.application.id]: action.payload.application
      };
    default:
      return state;
  }
};

export const ids = (state = new Set([]), action) => {
  switch (action.type) {
    case CATEGORIES_FETCH + SUCCESS:
      return new Set(
        action.payload.category.applications.reduce(
          applicationIdReduceFn, state
        )
      );
    case CATEGORIES_FETCH_ALL + SUCCESS:
      return new Set(
        action.payload.categories.reduce((accumulator, category) => (
          [
            ...accumulator,
            ...category.applications.reduce(applicationIdReduceFn, {})
          ]
        ), [])
      );
    default:
      return state;
  }
};

export const appProfile = (state = {}, action = {}) => {
  switch (action.type) {
    case APPLICATION_FETCH + SUCCESS:
      return action.payload.application;
    case APPLICATIONS_ADD_TO_FAVORITES + REQUEST:
    case APPLICATIONS_REMOVE_FROM_FAVORITES + REQUEST:
      return {
        ...state,
        inProgress: true
      };
    case APPLICATIONS_ADD_TO_FAVORITES + FAILURE:
    case APPLICATIONS_REMOVE_FROM_FAVORITES + FAILURE:
      return {
        ...state,
        inProgress: false
      };
    case APPLICATIONS_ADD_TO_FAVORITES + SUCCESS:
      return {
        ...state,
        inProgress: false,
        favorite: action.payload.favorite
      };
    case APPLICATIONS_REMOVE_FROM_FAVORITES + SUCCESS:
      return {
        ...state,
        inProgress: false,
        favorite: null
      };
    case REVIEW_CREATE + SUCCESS:
      return {
        ...state,
        review: action.payload.review.value
      };
    default:
      return state;
  }
};

export default combineReducers({
  ids,
  byId,
  appProfile
});
