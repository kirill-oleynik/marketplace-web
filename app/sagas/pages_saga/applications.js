import { all, put, take } from 'redux-saga/effects';

import { fetchAll } from '../../actions/categories_actions';
import {
  fetch, fetchApplicationGallery, fetchRating
} from '../../actions/applications_actions';
import {
  SUCCESS, FAILURE, APPLICATION_FETCH, APPLICATIONS_FETCH_GALLERY,
  APPLICATIONS_RATING_FETCH, CATEGORIES_FETCH_ALL
} from '../../constants';

export default function* loadApplicationsPage({ query }) {
  yield put(
    fetchAll()
  );

  yield put(
    fetch(query.slug)
  );

  yield put(
    fetchRating(query.slug)
  );

  yield put(
    fetchApplicationGallery(query.slug)
  );

  yield all([
    take([APPLICATION_FETCH + SUCCESS, APPLICATION_FETCH + FAILURE]),
    take([APPLICATIONS_FETCH_GALLERY + SUCCESS, APPLICATIONS_FETCH_GALLERY + FAILURE]),
    take([APPLICATIONS_RATING_FETCH + SUCCESS, APPLICATIONS_RATING_FETCH + FAILURE]),
    take([CATEGORIES_FETCH_ALL + SUCCESS, CATEGORIES_FETCH_ALL + FAILURE])
  ]);
}
