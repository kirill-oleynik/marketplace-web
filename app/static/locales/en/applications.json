{
  "applications": {
    "appProfile": {
      "requestDemo": "Reques Demo",
      "gallery": {
        "title": "Screenshots and Videos"
      },
      "description": {
        "title": "About",
        "readMore": "Read More",
        "readLess": "Read Less"
      },
      "extra": {
        "title": "Additional Information",
        "founded": "Founded",
        "email": "Email",
        "website": "Website",
        "phone": "Phone",
        "address": "Address"
      },
      "ratings": {
        "title": "Customer Ratings",
        "avarage": "Average Rating",
        "submit": "Click to Rate"
      },
      "favorites": {
        "add": "Add to Favorites",
        "added": "Added to Favorites",
        "remove": "Remove"
      }
    }
  }
}
