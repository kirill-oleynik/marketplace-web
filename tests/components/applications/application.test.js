import React from 'react';

import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { Application } from 'app/components/applications/application';

describe('#render', () => {
  test('it renders correctly', () => {
    const component = shallow(
      <Application
        canToggleFavorite
        deleteReview={() => {}}
        createReview={() => {}}
        addToFavorites={() => {}}
        removeFromFavorites={() => {}}
        t={(translation) => translation}
        gallery={{
          slides: []
        }}
        appProfile={{
          id: '1',
          slug: 'slug',
          title: 'title',
          description: 'description',
          summary: 'summary',
          phone: 'phone',
          email: 'email',
          website: 'website',
          address: 'address',
          founded: '2014-05-24',
          review: 3
        }}
        appRating={{
          average: 0,
          total: 0,
          votes: {
            1: 0,
            2: 0,
            3: 0,
            4: 0,
            5: 0
          }
        }}
      />
    );

    expect(
      toJSON(component)
    ).toMatchSnapshot();
  });
});
