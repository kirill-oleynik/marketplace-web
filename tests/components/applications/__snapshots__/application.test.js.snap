// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render it renders correctly 1`] = `
<div>
  <div
    className="d-flex flex-sm-wrap-down mb-50"
  >
    <div>
      <span
        className="app-item__img app-item__img--md mr-40"
        style={
          Object {
            "backgroundImage": "url('undefined')",
          }
        }
      />
    </div>
    <div
      className="d-flex flex-wrap flex-grow-1 justify-content-between pt-15"
    >
      <div
        className="mb-30"
      >
        <p
          className="font-24 font-700 mb-10"
        >
          title
        </p>
        <RatingPreview
          average={0}
          className="mb-30"
          t={[Function]}
          total={0}
        />
        <ToggleFavorite
          addToFavorites={[Function]}
          application={
            Object {
              "address": "address",
              "description": "description",
              "email": "email",
              "founded": "2014-05-24",
              "id": "1",
              "phone": "phone",
              "review": 3,
              "slug": "slug",
              "summary": "summary",
              "title": "title",
              "website": "website",
            }
          }
          disabled={false}
          removeFromFavorites={[Function]}
        />
      </div>
      <div>
        <MainButton
          className=""
          color="blue"
          size="lg"
          type="button"
        >
          appProfile.requestDemo
        </MainButton>
      </div>
    </div>
  </div>
  <div
    className="mb-20"
  >
    <p
      className="font-20 font-700 mb-30"
    >
      appProfile.gallery.title
    </p>
    <p
      className="font-20 font-700"
    >
      appProfile.description.title
    </p>
    <p
      className="in-black-050"
    >
      description
    </p>
  </div>
  <div
    className="divider divider--dark mb-30"
  />
  <div>
    <p
      className="font-20 font-700"
    >
      appProfile.ratings.title
    </p>
    <div
      className="d-flex flex-wrap justify-content-between"
    >
      <ApplicationRating
        data={
          Object {
            "average": 0,
            "total": 0,
            "votes": Object {
              "1": 0,
              "2": 0,
              "3": 0,
              "4": 0,
              "5": 0,
            },
          }
        }
        t={[Function]}
      />
      <div>
        <Translate(SubmitReview)
          onSubmit={[Function]}
          review={3}
        />
        <Translate(RemoveReview)
          application={
            Object {
              "address": "address",
              "description": "description",
              "email": "email",
              "founded": "2014-05-24",
              "id": "1",
              "phone": "phone",
              "review": 3,
              "slug": "slug",
              "summary": "summary",
              "title": "title",
              "website": "website",
            }
          }
          deleteReview={[Function]}
          review={3}
        />
      </div>
    </div>
  </div>
  <div
    className="divider divider--dark mb-30"
  />
  <div>
    <p
      className="font-20 font-700"
    >
      appProfile.extra.title
    </p>
    <Row
      className="mb-15"
      tag="div"
    >
      <Col
        md="4"
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <p
          className="font-16 font-700 mb-5"
        >
          appProfile.extra.founded
        </p>
        <p
          className="in-black-050"
        >
          May, 24 2014
        </p>
      </Col>
      <Col
        md="4"
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <p
          className="font-16 font-700 mb-5"
        >
          appProfile.extra.email
        </p>
        <a
          className="in-black-050"
          href="mailto:email"
        >
          email
        </a>
      </Col>
      <Col
        md="4"
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <p
          className="font-16 font-700 mb-5"
        >
          appProfile.extra.website
        </p>
        <a
          className="in-black-050"
          href="website"
          target="_blank"
        >
          website
        </a>
      </Col>
      <Col
        md="4"
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <p
          className="font-16 font-700 mb-5"
        >
          appProfile.extra.phone
        </p>
        <a
          className="in-black-050"
          href="tel:phone"
        >
          phone
        </a>
      </Col>
      <Col
        md="4"
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <p
          className="font-16 font-700 mb-5"
        >
          appProfile.extra.address
        </p>
        <p
          className="in-black-050"
        >
          address
        </p>
      </Col>
    </Row>
  </div>
</div>
`;
